Add custom object hooks

Simon-Campbell opened the pull request:
This adds a functionality for overriding/customising the structured object logged at request received/success/error. It is *like* the custom[Received|Success|Error]Message hooks, but works on customising the structured object.

It was opted not to overload *Message due to that being specifically about messages, and the risk that adjusting that could break an existing usage of that hook.

The README has been updated with example usage.

mcollina reviewed (changes requested):
Thanks for opening a PR! Can you please add a unit test?

Simon-Campbell commented:
> Thanks for opening a PR! Can you please add a unit test?

Thanks for feedback. I have now added tests as an extra commit.

mcollina commented on logger.js:
```diff
@@ -64,8 +64,13 @@ function pinoLogger (opts, stream) {
   const autoLoggingIgnore = opts.autoLogging && opts.autoLogging.ignore ? opts.autoLogging.ignore : null
   delete opts.autoLogging
 
+  const onRequestReceivedObject = opts.customReceivedObject && typeof opts.customReceivedObject === 'function' ? opts.customReceivedObject : undefined
```
Could you move the function to be a top-level function instead of a closure? I think the code would read better.

mcollina commented on logger.js:
```diff
@@ -147,9 +153,20 @@ function pinoLogger (opts, stream) {
+            req.log[level](receivedStringResult)
+          } else {
+            req.log[level](receivedObjectResult)
+          }
```
There are quite a a lot of conditionals in this block. Could you simplify them a bit and remove some duplications?

Simon-Campbell commented:
Hi @mcollina, 

I have addressed your feedback. I have also converted the common repeated pattern of (since I believe it further simplified what was being expressed.)

`const someEffectiveValue = opt.someOption && opt.someOption === 'function' ? opt.someOption : SomeDefaultOption;
`

to:

`
const someEffectiveValue = getFunctionOrDefault(opt.someOption, SomeDefaultOption);`

mcollina reviewed (approved):
lgtm
